src/main/java/hudson/plugins/gradle/injection/MavenExtensionsHandler.java

```diff
@@ -0,0 +1,93 @@
+
+public class MavenExtensionsHandler {
+
+    static final String LIB_DIR_PATH = "jenkins-gradle-plugin/lib";
+
+    private final MavenExtensionFileHandler geExtensionHandler = new MavenExtensionFileHandler(GRADLE_ENTERPRISE);
+    private final MavenExtensionFileHandler ccudExtensionHandler = new MavenExtensionFileHandler(CCUD);
+
+    public void copyGradleEnterpriseExtensionToAgent(FilePath rootPath) throws IOException, InterruptedException {
+        geExtensionHandler.copyExtensionToAgent(rootPath);
+    }
+
+    public FilePath getGradleEnterpriseExtensionPath(FilePath rootPath) throws IOException {
+        return geExtensionHandler.getAgentExtensionPath(rootPath);
+    }
+
+    public void copyCCUDExtensionToAgent(FilePath rootPath) throws IOException, InterruptedException {
+        ccudExtensionHandler.copyExtensionToAgent(rootPath);
+    }
+
+    public FilePath getCCUDExtensionPath(FilePath rootPath) throws IOException {
+        return ccudExtensionHandler.getAgentExtensionPath(rootPath);
+    }
+
+    public void deleteAllExtensionsFromAgent(FilePath rootPath) throws IOException, InterruptedException {
+        rootPath.child(LIB_DIR_PATH).deleteContents();
+    }
+
+    private static final class MavenExtensionFileHandler {
+        private final MavenExtension extension;
+
+        MavenExtensionFileHandler(MavenExtension extension) {
+            this.extension = extension;
+        }
+
+        public void copyExtensionToAgent(FilePath rootPath) throws IOException, InterruptedException {
+            copyResourceToNode(rootPath.child(LIB_DIR_PATH).child(extension.getJarName()), extension.getJarName());
+        }
+
+        public FilePath getAgentExtensionPath(FilePath rootPath) {
+            return rootPath.child(LIB_DIR_PATH).child(extension.getJarName());
+        }
+    }
+
+    public enum MavenExtension {
+        GRADLE_ENTERPRISE("gradle-enterprise-maven-extension"),
+        CCUD("common-custom-user-data-maven-extension");
+
+        final String name;
+        final Supplier<String> version;
+
+        public String getVersionFileName() {
+            return name + "-version.txt";
+        }
+
+        private String getExtensionVersion() {
+            try {
+                String resourceName = getVersionFileName();
+                try (InputStream version = MavenBuildScanInjection.class.getResourceAsStream("/versions/" + resourceName)) {
+                    if (version == null) {
+                        throw new IllegalStateException("Could not find resource: " + resourceName);
+                    }
+                    return IOUtils.toString(version, StandardCharsets.UTF_8);
+                }
+            } catch (IOException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        public String getJarName() {
+            return name + "-" + getExtensionVersion() + ".jar";
```

Comment:
I think we should be able to use `version.get()` here? Otherwise, we don't get much from caching, as we read the file again when the method is called.

Do we really need a version supplier here? Just for laziness? Maybe just read the version in the constructor and that's it?

Reply:
I missed, this, done. We shouldn't read it in the constructor as we should catch the exception only when we need to read it.